yt_dlp/extractor/generic.py

```diff
@@ -54,6 +54,7 @@ class GenericIE(InfoExtractor):
     _VALID_URL = r'.*'
     IE_NAME = 'generic'
     _NETRC_MACHINE = False  # Suppress username warning
+    visited_redirect_urls = set()
```

Comment:
making this a class variable will have unintended consequences, e.g. long-lived instances of ydl will raise the extractor error if revisiting a previously extracted URL

Reply:
For example, `yt-dlp URL1 URL1 -o %(autonumber)s.%(ext)s` should download twice. Similarly, if a playlist contains repeated item, both should download.

Reply:
Good catch. Confirmed the above example does trigger this. Since I'm only adding to the list during a redirect, I added a .clear() to the set on _real_extract() which should be called towards the beginning of a new extraction. This seems to correct the behavior, so that your example works.

Reply:
> I added a .clear() to the set on _real_extract() which should be called towards the beginning of a new extraction.

But when a redirect is followed, the generic extractor returns a URL result which is (most often) extracted again by the generic extractor, meaning `_real_extract()` is invoked again and the set will be cleared. So now this PR's changes don't really do anything at all

Reply:
To be clear: I'm not trying to just pick apart this PR or pick on you. This is not an easy problem to solve. I appreciate the effort, but this is not an acceptable solution. Nobody has found an acceptable solution yet, hence why the problem has gone unsolved for ~2.5 years.

Reply:
No worries, I appreciate the review. I'll continue to look at it, but I'll mark the PR as a draft in the meantime.

Reply:
I've looked at handling this in process_ie_result() in a more generic way. One of the issue is that this can also occur with infinite unique urls as well. Since we are executing recursively, I'm proposing instead to have a max depth limit to recursing through `url` `ie_result`s. I have branch with how that would look here: https://github.com/yt-dlp/yt-dlp/compare/master...adam-miller:yt-dlp:handle-infinite-redirects

I have an arbitrary limit set for the moment, but my thought was to make this a command line flag for consistent behavior, and either experimentally determine a "large" upper default limit, or make it infinite if not specified.